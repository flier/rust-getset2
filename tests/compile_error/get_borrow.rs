use std::sync::Arc;

use getset2::Getter;

#[derive(Getter)]
pub struct Struct {
    #[get(borrow)] // #[get(borrow)] is not allowed for a field that is not a Borrow<T>
    field: usize,
}

#[derive(Getter)]
#[get(borrow)] // #[get(borrow)] is ignored when it applied to the structure
pub struct Struct1 {
    field: usize,
}

#[derive(Getter)]
pub struct Struct2 {
    #[get(borrow(usize))]
    field: usize,
}

#[derive(Getter)]
pub struct Struct3 {
    #[get(borrow(isize))] // #[get(borrow)] is not allowed for a field that is not a Borrow<T>
    field: Arc<usize>,
}

#[derive(Getter)]
pub struct Struct4 {
    #[get(borrow_mut(isize))]
    // #[get(borrow)] is not allowed for a field that is not a BorrowMut<T>
    field: Arc<usize>,
}

fn main() {}
