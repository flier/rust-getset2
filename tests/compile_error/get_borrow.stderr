error: invalid #[get(..)] attribute, expected `borrow(...)`, found `borrow`
 --> tests/compile_error/get_borrow.rs:7:5
  |
7 |     #[get(borrow)] // #[get(borrow)] is not allowed for a field that is not a Borrow<T>
  |     ^

error: invalid #[get(..)] attribute, cannot find parameter `borrow` in this scope
  --> tests/compile_error/get_borrow.rs:12:1
   |
12 | #[get(borrow)] // #[get(borrow)] is ignored when it applied to the structure
   | ^

error[E0277]: the trait bound `Arc<usize>: Borrow<isize>` is not satisfied
  --> tests/compile_error/get_borrow.rs:25:5
   |
25 |       #[get(borrow(isize))] // #[get(borrow)] is not allowed for a field that is not a Borrow<T>
   |       ^ required by a bound introduced by this call
   |  _____|
   | |
26 | |     field: Arc<usize>,
   | |_________^ the trait `Borrow<isize>` is not implemented for `Arc<usize>`
   |
   = help: the trait `Borrow<T>` is implemented for `Arc<T>`

error[E0308]: mismatched types
  --> tests/compile_error/get_borrow.rs:31:5
   |
31 |       #[get(borrow_mut(isize))] // #[get(borrow)] is not allowed for a field that is not a Borrow<T>
   |  _____^________________-
   | |     |
   | |     expected `&mut isize`, found `&mut Arc<usize>`
32 | |     field: Arc<usize>,
   | |______________- expected `&mut isize` because of return type
   |
   = note: expected mutable reference `&mut isize`
              found mutable reference `&mut Arc<usize>`
